pkg/termination/parse.go

```diff
@@ -19,33 +19,69 @@ import (
 	"encoding/json"
 	"fmt"
 	"sort"
+	"time"
 
 	v1alpha1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// ParseMessage parses a termination message as results.
+// TerminationDetails stores information collected from a Pod status'
+// termination messages.
+type TerminationDetails struct {
+	// StartTimes is all the start times of containers in the Pod, in
+	// order, as reported by the container statuses' termination messages.
+	StartTimes []metav1.Time
+	// ResourceResults is all the resource results reported by containers
+	// in the Pod, in order, as reported by the container statuses'
+	// termination messages.
+	ResourceResults []v1alpha1.PipelineResourceResult
+}
+
+// ParseMessages parses all of a PodStatus' termination messages to produce a
+// list of PipelineResourceResults, and the start times for each step.
 //
-// If more than one item has the same key, only the latest is returned. Items
-// are sorted by their key.
-func ParseMessage(msg string) ([]v1alpha1.PipelineResourceResult, error) {
-	if msg == "" {
-		return nil, nil
-	}
-	var r []v1alpha1.PipelineResourceResult
-	if err := json.Unmarshal([]byte(msg), &r); err != nil {
-		return nil, fmt.Errorf("parsing message json: %v", err)
+// For ResourceResults, if more than one item has the same key for all
+// container statuses, only the last is preserved. Items are sorted by their
+// key.
+func ParseMessages(s corev1.PodStatus) (*TerminationDetails, error) {
+	var d TerminationDetails
+	for idx, cs := range s.ContainerStatuses {
+		if cs.State.Terminated != nil {
+			msg := cs.State.Terminated.Message
+			if msg == "" {
+				continue
+			}
+			var this []v1alpha1.PipelineResourceResult
+			if err := json.Unmarshal([]byte(msg), &this); err != nil {
+				return nil, fmt.Errorf("parsing message json for status %d: %v", idx, err)
+			}
+
+			for ridx, r := range this {
+				if r.Key == "StartedAt" {
+					t, err := time.Parse(time.RFC3339, r.Value)
+					if err != nil {
+						return nil, fmt.Errorf("parsing start time for status %d: %v", idx, err)
+					}
+					d.StartTimes = append(d.StartTimes, metav1.NewTime(t))
+					this = append(this[:ridx], this[ridx+1:]...) // Trim out this result.
```

Comment:
couple thoughts here:
1. we're modifying a list that we're also looping over.  even if go protects us here it might be worth approaching this differently because it can be eyebrow-raising.
2. we could just create a new list before all the looping starts and exclude this from being added to it.